pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,572 @@
+
+```ebnf {% showTitle=false %}
+Module
+    : Prelude "module" RelativeIdentifier
+    ;
+```
+
+If the module's identifier is [unscoped][identifier], this declares a _top-level module_.
+If the module's identifier is [scoped][identifier], this declares a _submodule_.
+
+Submodules are modules that are logically contained within another module (called the _parent_).
+Slice definitions within a submodule can reference types in parent modules without qualification.
+
+```slice
+module A // Declares a top-level module named `A`.
+struct AStruct {}
+```
+
+```slice
+module A::B // Declares a submodule named `B` that is contained within module `A`.
+typealias T = AStruct // Don't need to qualify as `A::AStruct`.
+```
+
+All slice definitions must be contained within modules - Slice definitions cannot exist at 'global' scope.
+If a Slice file contains no Slice definitions, it is legal to omit the module declaration.
+Otherwise, if a Slice file contains any Slice definitions, a module declaration is required,
+and any Slice definitions in the remainder of the file will be contained within the declared module.
+
+Module declarations must come before any Slice definitions and must come after any [file attributes][attribute] or [mode declarations][mode-declaration].
+There can be at most one module declaration per Slice file.
+
+```ebnf {% showTitle=false %}
+SliceFile
+    : SliceFilePrelude Module? Definition*
+    ;
+```
+
+It is legal for multiple Slice files to declare the same module.
+Slice definitions defined in those files will be in the same module, even if they're in different Slice files.
+
+For additional information on modules, see the [module][module-guide] page.
+
+### Primitive types
+
+Slice supports 17 primitive types (although `AnyClass` is only valid in [`Slice1`][compilation-mode-guide] mode).
+Primitive types are always available: they're defined in any scope of any Slice file.
+They can be referenced with their keywords:
+
+```ebnf {% showTitle=false %}
+Primitive
+    : "bool"      | "int8"  | "uint8"  | "int16"    | "uint16"    | "int32"   | "uint32"  | "varint32"
+    | "varuint32" | "int64" | "uint64" | "varint62" | "varuint62" | "float32" | "float64" | "string"
+    | "AnyClass"
+    ;
+```
+
+For additional information on primitive types, see the [primitive type][primitive-type-guide] page.
+
+### Collection types
+
+Slice supports 2 built-in collection types: sequences and dictionaries.
+Both types can be referenced with their respective keywords: `Sequence` and `Dictionary`,
+
+```ebnf {% showTitle=false %}
+Sequence
+    : "sequence" "<" TypeRef ">"
+    ;
+
+Dictionary
+    : "dictionary" "<" TypeRef "," TypeRef ">"
+    ;
+```
+
+Since they are are generic over the types of elements they contain, you must also specify their type arguments:
+
+```slice
+Sequence<bool>
+Dictionary<uint8, string>
+```
+
+Since dictionary keys must be comparable, there are restrictions on which types can be used for them.
```

Comment:
I didn't document it because we already talk about it in the language guide.

I've now added the restrictions to the reference, but should I leave the guide-version intact, or replace it with a link to this section of the reference?

Reply:
The Guide and Reference can occasionally duplicate information.

Reply:
Alright, then I'll leave the guide as-is.
Please look through my rewrite of the sequence/dictionary section then.